metalearn/metafeatures/base.py

```diff
@@ -0,0 +1,176 @@
+import inspect
+from abc import ABC, abstractmethod
+from typing import List, Callable, Dict, Union, Optional, Any
+import itertools
+
+from metalearn.metafeatures.constants import ProblemType, MetafeatureGroup
+
+class ResourceComputer:
+
+    def __init__(
+        self,
+        computer: Callable,
+        returns: List[str],
+        argmap: Optional[Dict[str,Any]] = {}
+    ) -> None:
+        """
+        Decorates ``computer``, a resource computing function
+        with metadata about that function.
+        
+        Parameters
+        ----------
+        computer
+            The function that computes the resources.
+        returns
+            The names of the resources that ``computer`` returns, specified in
+            the same order as ``computer`` returns them.
+        argmap
+            A custom map of ``computer``'s argument names to the global resource names
+            that will be passed as ``computer``'s arguments when ``computer`` is called.
+        """
+        self._computer = computer
+        self.returns = returns
+
+        self.argmap = {}
+
+        # reversing is needed because `self.defaults` gives the default
+        # argument values corresponding to the *last* `n` arguments in the
+        # function signature.
+        reversed_args = self.args[::-1]
+        reversed_defaults = self.defaults[::-1]
+        arg_default_pairs = itertools.zip_longest(reversed_args, reversed_defaults)
+
+        for local_name, default in arg_default_pairs:
+            # By default, just use the `computer` function's
+            # normal local argument names in the argmap,
+            # making sure to preserve default argument values
+            # when they are supplied.
+            if default is not None:
+                # The function has a default value for this arg;
+                # use that.
+                self.argmap[local_name] = default
+            else:
+                # This function has no default. Tell the system
+                # to pass in the global resource identified by
+                # this arg's ``local_name`` when calling this
+                # ``computer``.
+                self.argmap[local_name] = local_name
+
+        for local_name, resource_name in argmap.items():
+            # Now include any argument name or value overrides
+            # the developer has provided. Note: `resource_name`
+            # may be a global resource name (e.g. `"XSample"`) or
+            # a direct value for the argument (e.g. `5`)
+            self.argmap[local_name] = resource_name
+    
+    def __call__(self, *args, **kwargs):
+        """
+        Allows a ``ResourceComputer`` instance to be callable.
+        Just forwards all arguments on to self._computer.
+        """
+        return self._computer(*args, **kwargs)
```

Comment:
If we need a check like this, let's do it in the `__init__` so it only has to happen once.

Reply:
Here's my thoughts:

Since the API of `ResourceComputer` is developer-facing, meaning the values passed in to its `__init__` function will all be supplied by the developers, I'm happy with just relying on the errors raised by the system. I'm glad we validate all user-supplied values in the `Metafeatures` class `compute` function.

Concerning hard type checking, Maybe a nice way to do it would be to include a `mypy` type checking run in our Travis CI checks. That way we could rely on the typing info already present in our code to enforce correct typing, rather than having to add explicit checks in the code in addition to the typing info that's already there. I think that would be a separate issue though if that's something we're interested in.

Reply:
If you have time, let's add a mypy test.